src/linux/mod.rs

```diff
@@ -44,61 +48,119 @@ impl Sandbox for LinuxSandbox {
+///
+/// Returns PID of the child process if successful.
+fn spawn_sandbox_init(init_arg: ProcessInitArg, allow_networking: bool) -> Result<i32> {
+    unsafe {
+        // Initialize child process stack memory.
+        let stack_size = 1024 * 1024;
+        let child_stack = libc::mmap(
+            ptr::null_mut(),
+            stack_size,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_PRIVATE | libc::MAP_ANONYMOUS | libc::MAP_STACK,
+            -1,
+            0,
+        );
+        if child_stack == libc::MAP_FAILED {
+            return Err(IoError::last_os_error().into());
+        }
 
-        Ok(child)
+        // Stack grows downward on all relevant Linux processors.
+        let stack_top = child_stack.add(stack_size);
+
+        // Construct clone flags with required namespaces.
+        let mut flags =
+            libc::CLONE_NEWIPC | libc::CLONE_NEWNS | libc::CLONE_NEWPID | libc::CLONE_NEWUSER;
+        if !allow_networking {
+            flags |= libc::CLONE_NEWNET;
+        }
+
+        // Spawn sandbox init process.
+        let init_arg_raw = Box::into_raw(Box::new(init_arg)) as _;
+        let init_pid = libc::clone(sandbox_init, stack_top, flags | libc::SIGCHLD, init_arg_raw);
+        if init_pid == -1 {
+            Err(IoError::last_os_error().into())
+        } else {
+            Ok(init_pid)
+        }
     }
 }
 
-// NOTE: Since this new process is PID 1, it will be responsible for reaping all
-// orphans. We currently do not create a reaper for these and instead leak
-// zombies, relying on the spawned process being short-lived and not spawning a
-// lot of children.
-//
-/// Sandboxing steps executed in the new process' context.
-fn post_fork(uid: u32, gid: u32, allow_networking: bool) -> io::Result<()> {
-    // Create new procfs directory.
-    let new_proc_c = CString::new("/proc").unwrap();
-    namespaces::mount_proc(&new_proc_c)?;
+/// PID 1 inside the sandbox.
+///
+/// This function is the entry point for the process which is used to launch the
+/// sandboxee and act as init system for the PID namespace.
+extern "C" fn sandbox_init(arg: *mut libc::c_void) -> libc::c_int {
+    let init_arg: Box<ProcessInitArg> = unsafe { Box::from_raw(arg as _) };
+
+    match sandbox_init_inner(*init_arg) {
+        Ok(exit_code) => exit_code,
+        Err(err) => {
+            eprintln!("sandboxing failure: {err}");
+            1
+        },
+    }
+}
 
-    // Isolate networking using a network namespace.
-    if !allow_networking {
-        namespaces::create_user_namespace(0, 0, Namespaces::NETWORK)?;
+/// PID 1 inside the sandbox.
+///
+/// Wrapper to simplify error handling.
+fn sandbox_init_inner(mut init_arg: ProcessInitArg) -> io::Result<libc::c_int> {
+    // Hook up stdio to parent process.
+    rustix::io::dup2(stdio::stdin(), &mut init_arg.stdin)?;
+    rustix::io::dup2(stdio::stdout(), &mut init_arg.stdout)?;
+    rustix::io::dup2(stdio::stderr(), &mut init_arg.stderr)?;
```

Comment:
1. Isn't this closing the fds in `init_arg` and replacing them with copies from stdio? That seems backwards... It should be closing the current stdio and replacing it with `init_arg`, right?
2. If 1 is correct, `rustix` has safe wrappers for these.

```suggestion
    // Hook up stdio to parent process.
    rustix::io::dup2_stdin(init_arg.stdin)?;
    rustix::io::dup2_stdout(init_arg.stdout)?;
    rustix::io::dup2_stderr(init_arg.stderr)?;
```